source/common/common/utility.h

```diff
@@ -709,4 +709,34 @@ class InlineString : public InlineStorage {
   char data_[];
 };
 
+/**
+ * Utilities for working with containers.
+ */
+class Containers {
+public:
+  /**
+   * Remove elements that match the predicate from the list by swapping them to
+   * the end and truncating. Iterates over c once, order is not preserved.
```

Comment:
Which containers does this safely work on? Can we use https://en.cppreference.com/w/cpp/algorithm/remove? Would prefer to not repeat anything in STL already.

I like the way that cppreference says things like "ForwardIt must meet the requirements of LegacyForwardIterator." which makes 100% clear what is going on.

Reply:
it definitely works on list and vector; it would probably work on set too, but would be pretty pointless from a perf perspective.

Thinking about this, I bet std::remove_if would be adequate for this, and would have the possibly-desirable property of retaining relative order. I'm assuming std::remove_if would be O(n) though possibly with a higher constant factor if it winds up rebuilding the whole collection.

I should've thought of that before. Phil, do you want to try std::remove_if in your perf tests and see if it's reasonably close in performance to our new algo?

Reply:
FYI "why we need the custom iterator remover" -- it's mostly because we need to take further action on each removed item prior to removing it from the collection.

But std::remove_if looks like it does what we need in that respect (and the same predicates we are using now would work).

Reply:
std::remove_if is ~30% faster than the custom iterator-remover version at small and larger iteration counts. I'm chasing (confusing) test failures with it at the moment, but it looks better than writing this myself :).

and yes, the docs say that std::remove_if is O(n).

Reply:
nice! I was not expecting that. but STL algorithms have had a bit longer to mature and be tuned :)